#2065 Prevent saving invalid dates

CarlosNZ opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #2065

# 👩🏻‍💻 What does this PR do? 
Very small change -- just prevents the "onChange" method from running if there is a "validationError" from the mui DatePicker component.

# 🧪 How has/should this change been tested? 

Tested with the Date of Birth field for patients, but should be the same for any Date field that is using one of the MUI restriction props, i.e. `minDate`, `maxDate`, `disableFuture` or `disablePast`.

In this particular example, if the date is in the future, or before 1900, the "Save" button should not become enabled.

## 💌 Any notes for the reviewer?

> Might want to limit the year to `> 1000` or `> (current human age possible)` as well 🤔

The MUI DatePicker already defaults to a minDate of 01/01/1900, which would address this.

CarlosNZ commented on client/packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx:
```diff
@@ -20,19 +20,27 @@ export const BaseDatePickerInput: FC<
+  );
   const t = useTranslation('common');
 
-return (
+  return (
     <DatePicker
       slots={{
         textField: TextField,
       }}
-      onChange={(date,context) => {
-        const {validationError} = context;
+      onChange={(date, context) => {
+        const { validationError } = context;
 
-        setValidationError(validationError ? t(`error.date_${validationError}` as LocaleKey, {defaultValue: validationError}) : null);
-        onChange?.(date, context);
+        setValidationError(
+          validationError
+            ? t(`error.date_${validationError}` as LocaleKey, {
+                defaultValue: validationError,
+              })
+            : null
+        );
```
Sorry, most of this is just auto-formatting -- no actual changes.

CarlosNZ commented on client/packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx:
```diff
@@ -20,19 +20,27 @@ export const BaseDatePickerInput: FC<
+              })
+            : null
+        );
+        if (!validationError) onChange?.(date, context);
```
This one line is the only actual change.

roxy-dao commented:
Thanks Carl~

Would it become more complex if we disallow pressing `Ok` or `Save` buttons if the date is invalid? 🤔 Since seems a bit confusing to enter an invalid date, see that it's invalid, but you can still press ok or save

CarlosNZ commented:
> Would it become more complex if we disallow pressing `Ok` or `Save` buttons on the forms if the date is invalid? 🤔 Since seems a bit confusing to enter an invalid date, see that it's invalid, but you can still press ok or save

Isn't that what this PR does? Or am I misunderstanding you?

roxy-dao commented:
> > Would it become more complex if we disallow pressing `Ok` or `Save` buttons on the forms if the date is invalid? 🤔 Since seems a bit confusing to enter an invalid date, see that it's invalid, but you can still press ok or save
> 
> Isn't that what this PR does? Or am I misunderstanding you?

I can still click `Ok` or `Save`... it just doesn't save the date if it's invalid which ` if (!validationError) onChange?.(date, context);` is what this line is doing

CarlosNZ commented:
> I can still click `Ok` or `Save`... it just doesn't save the date if it's invalid which ` if (!validationError) onChange?.(date, context);` is what this line is doing

I guess you can still click save if you've also changed something else, but it shouldn't become enabled if the Invalid date is the only thing you've changed. 

I'll have to check how easy it would be to do what you're suggesting. Can't remember off the top of my head, will let you know.

mark-prins reviewed (changes requested):
This is what I get when I test:

<img width="1106" alt="Screenshot 2023-08-07 at 12 38 36 PM" src="https://github.com/openmsupply/open-msupply/assets/9192912/5513026a-157c-4bd1-b9c4-7da804ceb96f">

i.e. an invalid validation in the date picker does not result in save being disabled. 
It was disabling when using develop branch - but only for the case when the date is cleared with the delete / backspace key.

Also - the helper text doesn't show for the DOB control. 
And it would be good to add translations for `disableFuture` and `disablePast`

```
"error.date_disableFuture": "Date must not be in the future",
"error.date_disablePast": "Date must not be in the past",
```

or something like that.

CarlosNZ commented:
Okay @roxy-dao, @mark-prins -- have utilised the "useCustomError" hook and had to add an `onError` method to the BaseDatePicker so that JSONForms could be informed of the validation error.

9bd968af1

See inline comments for additional.

CarlosNZ commented on client/packages/programs/src/JsonForms/components/DateOfBirth.tsx:
```diff
@@ -80,9 +87,17 @@ const UIComponent = (props: ControlProps) => {
             value={dob ?? null}
             onChange={onChangeDoB}
             format="dd/MM/yyyy"
-            slotProps={{ textField: { sx: { width: 135 } } }}
+            width={135}
```
Took me ages to work out why I couldn't get the error text to show up -- then I realised the DOB component was overwriting the whole "slotProps" object! (That's a lot of props to wipe out just to set a `width` 😆)

roxy-dao reviewed (approved):
Thanks Carl~ 
Works perfectly~

The patient creation modal still allows invalid DOB, but it's not a required field there and shows up as empty on the patient detail view, so should be fine I guess

mark-prins commented:
sorry - found another edge case with the DoB component.
Changing the age sets the date value, but does not reset the validity:

<img width="450" alt="image" src="https://github.com/openmsupply/open-msupply/assets/9192912/3b90921f-b702-4f8a-900a-5e0d29ced907">

You can enter an invalid date, the date component shows the error, then you enter the age which updates the date with a valid value, but the error remains.

mark-prins commented:
The program enrolment forms also do not display the helper text

<img width="481" alt="image" src="https://github.com/openmsupply/open-msupply/assets/9192912/781b7e93-5d17-475d-b3b1-2b953c6f33a0">

or disable the form.. could these be updated too?

and the encounter enrolment form is a little broken.. If I select a valid date first try, then save is enabled

<img width="561" alt="image" src="https://github.com/openmsupply/open-msupply/assets/9192912/348be79d-e9aa-47fb-8970-cc330f9de235">

but if I enter an invalid date, then select a valid date, I'm unable to get save to re-enable!

mark-prins reviewed (commented):
Looking much better thanks! Still some scenarios that aren't quite working for me 😢

CarlosNZ commented on client/packages/programs/src/JsonForms/common/hooks/useJSONFormsCustomError.ts:
```diff
@@ -16,24 +16,30 @@ export const useJSONFormsCustomError = (
-          },
-        ])
-      );
+    const existingIndex = currentErrors.findIndex(
+      it => it.schemaPath === path && it.keyword === keyword
+    );
+    if (customError) {
+      if (existingIndex === -1)
+        dispatch(
+          Actions.updateErrors([
+            ...currentErrors,
+            {
+              instancePath: `/${path}`.replace(/\./g, '/'),
+              message: customError,
+              schemaPath: path,
+              keyword,
+              params: {},
+            },
+          ])
+        );
+      else {
+        const existingAction = currentErrors[existingIndex];
+        if (existingAction && existingAction?.message !== customError) {
+          existingAction.message = customError;
+          dispatch(Actions.updateErrors([...currentErrors]));
+        }
+      }
```
There was a problem in this implementation in that if the error message changed for a single element (e.g. goes from "Too small" to "Too big", this CustomError hook wouldn't get updated, since it's only checking for `path` and `keyword`.

So I've added an additional check to also check the messages are different and if so, replace the existing array element with the new message.

CarlosNZ commented on client/packages/programs/src/JsonForms/components/DateOfBirth.tsx:
```diff
@@ -60,6 +60,7 @@ const UIComponent = (props: ControlProps) => {
     setDoB(dob);
     handleChange(dobPath, dateFormatter(dob, 'yyyy-MM-dd'));
     handleChange(estimatedPath, true);
+    setCustomError(undefined);
```
Fixes the "Error not going away when manually changing the age" problem :)

CarlosNZ reviewed (commented):
Okay @mark-prins have addressed those problems (and dealt with a few more).

076629b00

See inline comments for more detail.

Turns out this was a bigger issue than originally thought!

There's just one thing I'm not totally happy with -- the way we handle errors when you *first* enter data is not consistent -- one of them nicely hides any error until you first leave the input field, the others all start complaining immediately. Let me know if you'd like to change *all* the date inputs to do it like this.

mark-prins commented on client/packages/common/src/ui/components/inputs/DatePickers/BaseDatePickerInput/BaseDatePickerInput.tsx:
```diff
@@ -17,22 +17,36 @@ export const BaseDatePickerInput: FC<
+          else
+            setInternalError(
+              validationError
+                ? t(`error.date_${validationError}` as LocaleKey, {
+                    defaultValue: validationError,
+                  })
```
with this block now used in three places it would be better to have it as a method - it was a bit of a hacky suggestion to start with sorry and I feel bad about it being used in diff places.
Would be better as a long form, less flexible approach so that we're not playing fast and loose with translation typings:

```
setInternalError(formattedDateError(t, validationError));
```

with a shared helper fn:

```
const formattedDateError = (t: TypedTFunction<LocaleKey>, error: string) => {
  if (!error) return null;

  switch (error) {
    case 'date_invalidDate':
      return t('error.date_invalidDate');
    case 'date_minDate':
      return t('error.date_minDate');
    case 'date_maxDate':
      return t('error.date_maxDate');
    case 'date_disablePast':
      return t('error.date_disablePast');
    case 'date_disableFuture':
      return t('error.date_disableFuture');
    default:
      return error;
  }
};
```

mark-prins reviewed (commented):
The program enrolment is working nicely thanks! showing the error and enforcing validation.
Am getting half of that with the encounter enrolment - the validation is fine but there's no error messaging:

<img width="529" alt="image" src="https://github.com/openmsupply/open-msupply/assets/9192912/4b1a5abc-6c5c-4ca4-b6e2-089f13511f2b">

Agree with you - would be nice to have the initial entry consistent. I'm getting this when creating a new stocktake:

<img width="379" alt="image" src="https://github.com/openmsupply/open-msupply/assets/9192912/f66d17f8-0b00-475e-906a-9a3c53ec1ed4">

I've just started entering a data and it is complaining already 🙄 
Not sure if we can do anything with that though, no way of knowing that the user has finished entering the whole date.
More important to change, is that the component isn't using the translations:

<img width="381" alt="image" src="https://github.com/openmsupply/open-msupply/assets/9192912/3f6db117-51af-4842-8a64-f8c13652c48d">

CarlosNZ commented:
Okay, I think I've caught all cases now, @mark-prins

5003a606d

- Checked on Patient birthdate, Program enrolments (both date fields), New Encounter modal, Stocktake date
- Removed "hasData" check from "Date.tsx" and made into a general "don't show error on initial entry" functionality in "BaseDatePickerInput" (so you shouldn't see any error until after initial `onBlur` event, or after selecting from Date picker -- after that, errors appear immediately)
- added generic `getTranslatedDateError` method in BaseDatePickerInput and removed repetition of this construction

mark-prins reviewed (approved):
🎉  thanks for all the additions - and your patience!
The date pickers are so much better now!
